sklearn/utils/_encode.py

```diff
@@ -118,26 +206,66 @@ def _check_unknown(values, known_values, return_mask=False):
         Additionally returned if ``return_mask=True``.
 
     """
-    if values.dtype == object:
+    valid_mask = None
+
+    if values.dtype.kind in 'UO':
+        values_set = set(values)
+        values_set, nan_in_values = _extract_missing(values_set,
+                                                     none_is_missing=False)
+
         uniques_set = set(known_values)
-        diff = list(set(values) - uniques_set)
+        uniques_set, nan_in_uniques = _extract_missing(uniques_set,
+                                                       none_is_missing=False)
+        diff = values_set - uniques_set
+        is_missing_in_uniques = bool(nan_in_uniques)
+
+        nan_in_diff = nan_in_values and not nan_in_uniques
+
+        def is_valid(value):
+            value_in_set = value in uniques_set
+            if nan_in_uniques and is_scalar_nan(value):
```

Comment:
This is confusing. It's performing `if bool(nan_in_uniques) and ...` but `bool(nan_in_uniques) == is_missing_in_uniques` and we return the latter.

I think the logic is not right here.

I also thing we should fast path the case where value_in_set.

See suggestion below